### PR TITLE
Implement storage support for CodeChat app in applab.

### PR DESCRIPTION
**Pull request by davidsbailey**
This adds support for basic CodeChat app:
1. Implement rudimentary data browser UI on top of forms API.
2. Implement Create and Read storage operations on top of v2 forms API. 
3. Add ability to set and get javascript attributes on HTML elements.

The data is structured as outlined in the CSP Data Storage Proposal doc (https://docs.google.com/a/code.org/document/d/1suoZsC0pbRKk6NYd0sTj6tvDNOy-yipbE-D9iES3CWY/edit#heading=h.b4qa11ojnntv):
1. a code.org account has a collection of apps (not yet implemented)
2. an app has a collection of tables
3. a table has a collection of records
4. a record has a collection of key-value pair fields, all stored in a single jSON string

heirarchy

Future work (incomplete list!):
- Search, Update and Delete operations
- schema enforcement (requiring a given field key to always have the same value type within a given table)
- associating a unique "app id" with each applab "app"
- simple key-value pair get/set operations
- login integration
- Same Origin Policy workaround for localhost development

### CODE REVIEW

**Comment by davidsbailey** [comment]
// CodeChat source code

createHtmlBlock('nameTitle', 'Welcome to CodeChat! Please enter your name:<br>');
createTextInput('nameInput', '');
createButton('nameButton', 'Save');
var name;
attachEventHandler('nameButton', 'click', render);
attachEventHandler('nameInput', 'keypress', function(e) {
  if (e.keyCode === 13) {
    render();
  }
});
function render() {
  name = getText('nameInput');
  deleteHtmlBlock('nameTitle');
  deleteHtmlBlock('nameInput');
  deleteHtmlBlock('nameButton');

  createHtmlBlock('id', 'Welcome to CodeChat, ' + name + '!');
  createHtmlBlock('chatText', '');
  setStyle('chatText', 'overflow: scroll; border-width: 1px; border-style: solid; width: 375px; height: 200px;');

  createButton('refreshButton', 'refresh');
  createHtmlBlock('id', '<br>');
  attachEventHandler('refreshButton', 'click', refresh);
  
  createTextInput('messageInput', '');
  createButton('sendButton', 'send');
  attachEventHandler('sendButton', 'click', send);
  attachEventHandler('messageInput', 'keypress', function(e) {
    if (e.keyCode === 13) {
      send();
    }
  });
  
  refresh();
}
function refresh() { 
  readRecords({tableName: 'ChatMessage'}, function(records) {
    var text = '';
    for (var i = 0; i < records.length; i++) {
      text += records[i].name + ' ';
      var time = (new Date(records[i].time)).toLocaleTimeString();
      text += '[' + time + ']: ';
      text += records[i].message + '\n';  
    }
    setText('chatText', text);
    // scroll to the bottom of the chat window.
    var scrollHeight = getAttribute('chatText', 'scrollHeight');
    setAttribute('chatText', 'scrollTop', scrollHeight);
  });
}
function send() {
  var time = Date.now();
  var message = getText('messageInput');
  setText('messageInput', '');
  createRecord({tableName: 'ChatMessage', time:time, name:name, message:message}, function(record) { 
    refresh();
  });
}

**Comment by davidsbailey** [comment]
bigger view of table editor portion of "data browser":
edit-csp-table

File: apps/src/webapp/extraControlRows.html.ejs [hunk 1]
```diff
@@ -19,6 +19,12 @@
   </div>
 <% } %>
 
+<div id="view-data-cell" class="view-data-cell-none">
```
Comment (cpirich) [inline apps/src/webapp/extraControlRows.html.ejs]: I think this should be inside the if (debugButtons) or a similar conditional. I think when config.hide_source is true, we don't want to show just this button.

File: apps/src/webapp/formStorage.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  req.setRequestHeader("Content-Type", "application/json; charset=UTF-8");
+  req.send(JSON.stringify(postData));
+};
+
+var handlePutRecord = function(record, callback, tableSecret) {
+  if (this.readyState !== 4) {
+    return;
+  }
+  if (this.status !== 201) {
+    console.log('unexpected http status ' + this.status);
```
Comment (cpirich) [inline apps/src/webapp/formStorage.js]: It seems like the callback should be called in the event of an async error so that the user's program doesn't hang indefinitely with no error. perhaps the callback needs to have a status code in addition to the record?

File: apps/src/webapp/formStorage.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+      tableName, 
+      putRecord.bind(this, record, callback));
+};
+
+var putRecord = function(record, callback, tableSecret) {
+  var req = new XMLHttpRequest();
+  req.onreadystatechange = handlePutRecord.bind(req, record, callback, tableSecret);
+  var url = '//' + getFormDataHost() + '/v2/forms/CspTable/' + tableSecret +
+      '/children/CspRecord';
+  delete record.tableName;
```
Comment (cpirich) [inline apps/src/webapp/formStorage.js]: why is this deleted? Is this temporary?

File: apps/src/webapp/formStorage.js [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+  var postData = {record_data_s: JSON.stringify(record)};
+  req.open('POST', url);
```
Comment (cpirich) [inline apps/src/webapp/formStorage.js]: third parameter (async) is missing - I assume it should be set to true since you are supplying an onreadystatechange function

File: apps/src/webapp/formStorage.js [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+  req.open('GET', url, true);
+  req.send();
+};
+
+var handleFetchRecords = function(tableName, searchParams, callback) {
+  if (this.readyState !== 4) {
+    return;
+  }
+  if (this.status !== 200) {
+    console.log('readRecords failed with status ' + this.status);
```
Comment (cpirich) [inline apps/src/webapp/formStorage.js]: same comment as above - how to communicate async failures?

File: apps/src/webapp/levels.js [hunk 6]
```diff
@@ -42,6 +42,8 @@ levels.ec_simple = {
     {'func': 'setParent', 'title': 'Set an element to become a child of a parent element', 'category': 'General', 'params': ["'id'", "'parentId'"] },
     {'func': 'setPosition', 'title': 'Position an element with x, y, width, and height coordinates', 'category': 'General', 'params': ["'id'", "0", "0", "100", "100"] },
     {'func': 'setStyle', 'title': 'Add CSS style text to an element', 'category': 'General', 'params': ["'id'", "'color:red;'"] },
+    {'func': 'getAttribute', 'category': 'General', 'params': ["'id'", "'innerHTML'"], 'type': 'value' },
```
Comment (cpirich) [inline apps/src/webapp/levels.js]: rather than innerHTML as the example attribute name, can you think of something more common? Maybe scrollHeight like you had to use in your sample. My worry is that we encourage students to party on innerHTML, but maybe they will need to do that.

File: apps/src/webapp/webapp.js [hunk 7]
```diff
@@ -505,6 +506,14 @@ Webapp.init = function(config) {
     }
   }
 
+  var viewDataButton = document.getElementById('viewDataButton');
+  dom.addClickTouchEvent(viewDataButton, Webapp.onViewData);
+
+  if (!StudioApp.hideSource) {
```
Comment (cpirich) [inline apps/src/webapp/webapp.js]: If you only added this button when the debugButtons flag was set (per my earlier comment), you could rewrite this a little.

File: apps/src/webapp/webapp.js [hunk 8]
```diff
@@ -1328,6 +1345,30 @@ Webapp.createDropdown = function (opts) {
...
+
+// Whitelist of HTML Element attributes which can be modified, to
+// prevent DOM manipulation which would violate the sandbox.
+Webapp.mutableAttributes = ['innerHTML', 'scrollTop'];
+
+Webapp.setAttribute = function (opts) {
+  var divWebapp = document.getElementById('divWebapp');
+  var element = document.getElementById(opts.elementId);
+  var attribute = String(opts.attribute);
+  var value = String(opts.value);
```
Comment (cpirich) [inline apps/src/webapp/webapp.js]: independent of my comment above, should all attribute values be strings? is it possible that some should be numbers (like IMG width, height)? If you could just use opts.value directly, it might be better.

File: apps/src/webapp/webapp.js [hunk 9]
```diff
@@ -1556,6 +1597,32 @@ Webapp.clearTimeout = function (opts) {
   window.clearTimeout(opts.timeoutId);
 };
 
+Webapp.createRecord = function (opts) {
+  var record = codegen.marshalInterpreterToNative(opts.record);
+  FormStorage.createRecord(record,
+      Webapp.handleCreateRecord.bind(this, opts.callback));
```
Comment (cpirich) [inline apps/src/webapp/webapp.js]: any synchronous return value needed for success/failure? (applies to this and readRecords below)

File: apps/src/webapp/webapp.js [hunk 10]
```diff
@@ -1618,6 +1685,28 @@ var checkFinished = function () {
   return false;
 };
 
+// TODO(dave): move this logic to dashboard.
```
Comment (cpirich) [inline apps/src/webapp/webapp.js]: agree, I think dashboard should pass the URL to the webapp code in the config block. and the webapp code should fail gracefully if it is not passed in (or if it is being run without the dashboard).

**Comment by cpirich** [comment]
All my comments are in. One general comment: use of console.log is problematic because it causes problems in IE9. So you should decide if you want to remove those or substitute something else that will just write to the webapp error console.

I didn't review the pegasus changes (not my area of expertise).

Overall, I'm super happy that you've got this in! I'm ok if you want to merge now and handle all of the feedback/comments afterwards.

File: apps/src/webapp/formStorage.js (hunk 3 above)
Comment (davidsbailey) [inline apps/src/webapp/formStorage.js]: record.tableName is deleted so that it does not show up in the data browser UI. Alternately, I could write this field to storage and simply remove it before displaying it data browser UI.

File: apps/src/webapp/formStorage.js (hunk 2 above)
Comment (davidsbailey) [inline apps/src/webapp/formStorage.js]: Let's discuss in person. Recovering from storage errors sounds advanced -- if we do call back to user code on failure, one possibility is a separate, optional onFailure handler for which we provide a simple default implementation.

File: apps/src/webapp/extraControlRows.html.ejs (hunk 1 above)
Comment (davidsbailey) [inline apps/src/webapp/extraControlRows.html.ejs]: Done

File: apps/src/webapp/formStorage.js (hunk 4 above)
Comment (davidsbailey) [inline apps/src/webapp/formStorage.js]: done

File: apps/src/webapp/levels.js (hunk 6 above)
Comment (davidsbailey) [inline apps/src/webapp/levels.js]: changed to scrollHeight

File: apps/src/webapp/webapp.js (hunk 7 above)
Comment (davidsbailey) [inline apps/src/webapp/webapp.js]: done

File: apps/src/webapp/webapp.js (hunk 8 above)
Comment (davidsbailey) [inline apps/src/webapp/webapp.js]: done.

File: apps/src/webapp/webapp.js (hunk 9 above)
Comment (davidsbailey) [inline apps/src/webapp/webapp.js]: leaving as is for now, let's discuss this in person.

**Comment by davidsbailey** [comment]
Thanks for the feedback, Chris! I've addressed some comments and added TODOs for others. Merging now to get this in time for a demon on Monday.
